### PR TITLE
Add automatic generation of CMakeLists.txt for msg packages

### PR DESCRIPTION
**Pull request by jpbusch**
- automatically generate `CMakeLists.txt` for message packages
  - implemented in `asn1ToRosMsg.py`
- only add necessary msg files to CMakeLists
- move some access functions for dedicated msg types, to avoid unecessary dependencies

### CODE REVIEW

**Comment by jpbusch** [comment]
Open questions:
- [ ] also remove non-build msg files? Would be a 3-liner.
- [ ] if removing the msgs, should we also remove the corresponding coding and conversion files?
- [ ] generate the CMakeLists.txt with jinja? Disadvantage: required jinja package

**Comment by lreiher** [comment]
> Open questions:
> 
> * [ ]  also remove non-build msg files? Would be a 3-liner.

Yes.

> * [ ]  if removing the msgs, should we also remove the corresponding coding and conversion files?

Conversion doesn't make sense if ROS messages are removed, so should be removed as well. I tend to keep coding though as the ground-truth asn1c output.

> * [ ]  generate the CMakeLists.txt with jinja? Disadvantage: required jinja package

At the current point, I don't think it's necessary. Would replace the multiple `f.write` calls with a single multi-line text though.

File: etsi_its_msgs_utils/include/etsi_its_msgs_utils/impl/cam/cam_getters.h
Comment (lreiher) [inline etsi_its_msgs_utils/include/etsi_its_msgs_utils/impl/cam/cam_getters.h:1]: Why are there these additions in cam and cdd access functions?

File: utils/codegen/codegen-rust/asn1ToRosMsg.py
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:1]: not yet reviewed

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 1]
```diff
@@ -142,6 +219,26 @@ def main():
...
+    elif args.type == "vam_ts":
+        msg_type = "VAM"
+
+    msg_files = [msg_type]
+    generate_required_msgs(os.path.join(args.output_dir, f"{msg_type}.msg"), msg_files)
+    msg_files.sort()
+    generate_cmakelists(msg_files, os.path.join(args.output_dir, "../CMakeLists.txt"), args.type)
+    
+    for f in glob.glob(os.path.join(args.output_dir, "*.msg")):
+        if os.path.basename(f).split('.')[0] not in msg_files:
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:239]:
```suggestion
        if os.path.splitext(f)[0] not in msg_files:
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 2]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
 
     return asn1_raw
 
+def generate_required_msgs(parent_file_path: str, file_list: list = []) -> list:
+
+    # load contents of msg file
+    with open(parent_file_path, 'r') as file:
+        lines = file.readlines()
+        for line in lines:
+            # if line doesnt starts with # or empty line, get first word
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:102]:
```suggestion
            # if line doesnt start with # or empty line, get first word
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 3]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
...
+            if not line.startswith("#") and line.strip() != "":
+                msg_type = line.split()[0]
+                # if ends with message type ends with [], remove []
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:105]:
```suggestion
                # if message type ends with [], remove []
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 4]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
...
+                msg_type = msg_type[:-2] if msg_type.endswith("[]") else msg_type
+                if msg_type not in file_list and os.path.isfile(f"{os.path.dirname(parent_file_path)}/{msg_type}.msg"):
+                    file_list.append(msg_type)
+                    generate_required_msgs(f"{os.path.dirname(parent_file_path)}/{msg_type}.msg", file_list)
+    
+    # make sure there are no duplicates and sort alphabetically
+    file_list = list(set(file_list))
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:112]:
```suggestion
    file_list = sorted(list(set(file_list)))
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 5]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
...
+
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:113]:
```suggestion
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 6]
```diff
@@ -142,6 +219,26 @@ def main():
...
+    if args.type == "cpm_ts":
+        msg_type = "CollectivePerceptionMessage"
+    elif args.type == "cam_ts":
+        msg_type = "CAM"
+    elif args.type == "vam_ts":
+        msg_type = "VAM"
+
+    msg_files = [msg_type]
+    generate_required_msgs(os.path.join(args.output_dir, f"{msg_type}.msg"), msg_files)
+    msg_files.sort()
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:235]:
```suggestion
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 7]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
...
+        lines = file.readlines()
+        for line in lines:
+            # if line doesnt starts with # or empty line, get first word
+            if not line.startswith("#") and line.strip() != "":
+                msg_type = line.split()[0]
+                # if ends with message type ends with [], remove []
+                msg_type = msg_type[:-2] if msg_type.endswith("[]") else msg_type
+                if msg_type not in file_list and os.path.isfile(f"{os.path.dirname(parent_file_path)}/{msg_type}.msg"):
+                    file_list.append(msg_type)
+                    generate_required_msgs(f"{os.path.dirname(parent_file_path)}/{msg_type}.msg", file_list)
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:109]:
```suggestion
                    file_list = generate_required_msgs(f"{os.path.dirname(parent_file_path)}/{msg_type}.msg", file_list)
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 8]
```diff
@@ -142,6 +219,26 @@ def main():
...
+    # handle special cases
+    if args.type == "cpm_ts":
+        msg_type = "CollectivePerceptionMessage"
+    elif args.type == "cam_ts":
+        msg_type = "CAM"
+    elif args.type == "vam_ts":
+        msg_type = "VAM"
+
+    msg_files = [msg_type]
+    generate_required_msgs(os.path.join(args.output_dir, f"{msg_type}.msg"), msg_files)
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:234]:
```suggestion
    msg_files = generate_required_msgs(os.path.join(args.output_dir, f"{msg_type}.msg"), [msg_type])
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 9]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
 
     return asn1_raw
 
+def generate_required_msgs(parent_file_path: str, file_list: list = []) -> list:
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:96]:
```suggestion
def findDependenciesOfRosMessageType(parent_file_path: str, file_list: List[str] = []) -> List[str]:
```

File: utils/codegen/codegen-rust/asn1ToRosMsg.py [hunk 10]
```diff
@@ -92,6 +93,82 @@ def asn1Definitions(files: List[str]) -> Dict[str, str]:
...
+                    file_list.append(msg_type)
+                    generate_required_msgs(f"{os.path.dirname(parent_file_path)}/{msg_type}.msg", file_list)
+    
+    # make sure there are no duplicates and sort alphabetically
+    file_list = list(set(file_list))
+    
+    
+    return file_list
+
+def generate_cmakelists(msg_files: list, file_path: str, type: str) -> None:
```
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToRosMsg.py:117]:
```suggestion
def generateCMakeLists(msg_files: list, file_path: str, type: str) -> None:
```

File: utils/codegen/codegen-rust/asn1ToConversionHeader.py
Comment (lreiher) [inline utils/codegen/codegen-rust/asn1ToConversionHeader.py:1]: see comments in asn1ToRosMsg.py

**Pull request merged**